pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,395 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package policycell
+
+import (
+	"context"
+	"log/slog"
+	"net/netip"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/cilium/stream"
+
+	"github.com/cilium/cilium/pkg/container/set"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	ipcachetypes "github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/monitor/agent"
+	monitorapi "github.com/cilium/cilium/pkg/monitor/api"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	policytypes "github.com/cilium/cilium/pkg/policy/types"
+	"github.com/cilium/cilium/pkg/source"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+type PolicyImporter interface {
+	UpdatePolicy(*policytypes.PolicyUpdate)
+}
+
+type policyImporterParams struct {
+	cell.In
+
+	Log      *slog.Logger
+	JobGroup job.Group
+
+	Repo            policy.PolicyRepository
+	EndpointManager endpointmanager.EndpointManager
+	IPCache         *ipcache.IPCache
+	MonitorAgent    agent.Agent
+}
+
+type policyImporter struct {
+	log          *slog.Logger
+	repo         policy.PolicyRepository
+	epm          epmanager
+	ipc          ipcacher
+	monitorAgent agent.Agent
+
+	// prefixesByResources is the list of prefixes
+	// that belong to each resource. This is tracked separately
+	// so we can allocate and release prefixes as policy changes.
+	prefixesByResource map[ipcachetypes.ResourceID][]netip.Prefix
+
+	q chan *policytypes.PolicyUpdate
+}
+
+type ipcacher interface {
+	UpsertMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	RemoveMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	WaitForRevision(ctx context.Context, rev uint64) error
+}
+
+type epmanager interface {
+	UpdatePolicy(idsToRegen *set.Set[identity.NumericIdentity], fromRev, toRev uint64)
+}
+
+func newPolicyImporter(cfg policyImporterParams) PolicyImporter {
+	i := &policyImporter{
+		log:          cfg.Log,
+		repo:         cfg.Repo,
+		epm:          cfg.EndpointManager,
+		ipc:          cfg.IPCache,
+		monitorAgent: cfg.MonitorAgent,
+
+		q: make(chan *policytypes.PolicyUpdate, option.Config.PolicyQueueSize),
+
+		prefixesByResource: map[ipcachetypes.ResourceID][]netip.Prefix{},
+	}
+
+	buf := stream.Buffer(
+		stream.FromChannel(i.q),
+		option.Config.PolicyQueueSize, 10*time.Millisecond,
```

Comment:
Looks like we could move the policy queue size option from `option.DaemonConfig` to a cell-specific flag and avoid a dependency on `pkg/option` while at it (or as a follow-up), as it is only used in this cell.

Reply:
great idea!